### PR TITLE
arrow_cast to cast to timestamptz

### LINKED ISSUE #5913: allow `ARROW_CAST` to cast to `TimestampTz`
### Is your feature request related to a problem or challenge?

`ARROW_CAST` was added in #5166  by @alamb which enable casting to arbitrary arrotypes.

However, casting to timetamptz (e.g. `Timestamp(Nanosecond, Some("+00:00"))` isn't supported yet

### Describe the solution you'd like

make this work
```
❯ select arrow_cast(now(), 'Timestamp(Nanosecond, Some( "+08:00" ))');
+----------------------------------+
| now()                            |
+----------------------------------+
| 2023-04-07T19:42:22.197975+08:00 |
+----------------------------------+
1 row in set. Query took 0.002 seconds.
```

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by waitingkuo**
# Which issue does this PR close?

Closes #5913

cc @alamb 

# Rationale for this change

# What changes are included in this PR?

allow arrow_cast to cast to timestamptz

```bash
❯ select arrow_cast(timestamp '2000-01-01T00:00:00', 'Timestamp(Nanosecond, Some( "+08:00" ))');
+-----------------------------+
| Utf8("2000-01-01T00:00:00") |
+-----------------------------+
| 2000-01-01T08:00:00+08:00   |
+-----------------------------+
1 row in set. Query took 0.010 seconds.
```
```bash
❯ select arrow_cast(timestamptz '2000-01-01T00:00:00Z', 'Timestamp(Nanosecond, Some( "+08:00" ))');
+------------------------------+
| Utf8("2000-01-01T00:00:00Z") |
+------------------------------+
| 2000-01-01T08:00:00+08:00    |
+------------------------------+
1 row in set. Query took 0.002 seconds.
```
```bash
❯ select arrow_cast(0, 'Timestamp(Nanosecond, Some( "+08:00" ))');
+---------------------------+
| Int64(0)                  |
+---------------------------+
| 1970-01-01T08:00:00+08:00 |
+---------------------------+
1 row in set. Query took 0.004 seconds.
```

# Are these changes tested?

yes

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/sql/src/expr/arrow_cast.rs [hunk 1]
```diff
@@ -167,6 +167,34 @@ impl<'a> Parser<'a> {
...
+                let timezone = self.parse_double_quoted_string("Timezone")?;
+                self.expect_token(Token::RParen)?;
+                Ok(Some(timezone))
+            }
+            tok => Err(make_error(
+                self.val,
+                &format!("finding Timezone for {context}, got {tok}"),
+            )),
+        }
+    }
```
Comment (waitingkuo) [inline datafusion/sql/src/expr/arrow_cast.rs:185]: add a parser to parse timezone `Option<String>`
either `None`, or `Some("SomeTimezone")`

File: datafusion/sql/src/expr/arrow_cast.rs [hunk 2]
```diff
@@ -167,6 +167,34 @@ impl<'a> Parser<'a> {
...
+    fn parse_double_quoted_string(&mut self, context: &str) -> Result<String> {
+        match self.next_token()? {
+            Token::DoubleQuotedString(s) => Ok(s),
+            tok => Err(make_error(
+                self.val,
+                &format!("finding double quoted string for {context}, got '{tok}'"),
+            )),
+        }
+    }
+
```
Comment (waitingkuo) [inline datafusion/sql/src/expr/arrow_cast.rs:197]: add parser to parse the double quoted string e.g. "+00:00"

File: datafusion/sql/src/expr/arrow_cast.rs [hunk 3]
```diff
@@ -580,8 +632,15 @@ mod test {
...
-            //DataType::Timestamp(TimeUnit::Nanosecond, Some("UTC".into())),
+            // we can't cover all possible timezones, here we only test utc and +08:00
+            DataType::Timestamp(TimeUnit::Nanosecond, Some("+00:00".into())),
+            DataType::Timestamp(TimeUnit::Microsecond, Some("+00:00".into())),
+            DataType::Timestamp(TimeUnit::Millisecond, Some("+00:00".into())),
+            DataType::Timestamp(TimeUnit::Second, Some("+00:00".into())),
+            DataType::Timestamp(TimeUnit::Nanosecond, Some("+08:00".into())),
+            DataType::Timestamp(TimeUnit::Microsecond, Some("+08:00".into())),
+            DataType::Timestamp(TimeUnit::Millisecond, Some("+08:00".into())),
+            DataType::Timestamp(TimeUnit::Second, Some("+08:00".into())),
```
Comment (alamb) [inline datafusion/sql/src/expr/arrow_cast.rs:643]: Some other test cases that would be good:

`""` (empty timezone)

Also errors:

Three double quotes
```
"Timestamp(Nanosecond, Some("+00:00""))" 
```

**Comment by tustvold** [comment]
```
❯ select arrow_cast(timestamp '2000-01-01T00:00:00', 'Timestamp(Nanosecond, Some( "+08:00" ))'); +-----------------------------+ | Utf8("2000-01-01T00:00:00") | +-----------------------------+ | 2000-01-01T08:00:00+08:00 | +-----------------------------+ 1 row in set. Query took 0.010 seconds.
```

IMO this behaviour is a bug in arrow-cast that should be fixed, I'm not sure if we wish to expose this.

The "correct" behaviour would be

```
❯ select arrow_cast(timestamp '2000-01-01T00:00:00', 'Timestamp(Nanosecond, Some( "+08:00" ))'); +-----------------------------+ | Utf8("2000-01-01T00:00:00") | +-----------------------------+ | 2000-01-01T00:00:00+08:00 | +-----------------------------+ 1 row in set. Query took 0.010 seconds.
```

**Review by alamb — APPROVED** [review]
Looks great @waitingkuo  -- I had some small suggestions on some other tests but we can also add them as a follow on PR

**Comment by tustvold** [comment]
https://github.com/apache/arrow-rs/pull/4034 should fix the timestamp handling in arrow-cast, if we are quick it may still make the release this afternoon

File: datafusion/sql/src/expr/arrow_cast.rs (hunk 3 above)
Comment (waitingkuo) [reply datafusion/sql/src/expr/arrow_cast.rs:659]: thanks @alamb , more error handling and test cases added

File: datafusion/sql/src/expr/arrow_cast.rs [hunk 4]
```diff
@@ -673,10 +748,21 @@ mod test {
...
+                r#"Timestamp(Nanosecond, Some("+00:00))"#,
+                r#"parsing "+00:00 as double quoted string: last char must be ""#,
+            ),
+            (
+                r#"Timestamp(Nanosecond, Some(""))"#,
+                r#"parsing "" as double quoted string: empty string isn't supported"#,
+            ),
+            (
+                r#"Timestamp(Nanosecond, Some("+00:00""))"#,
+                r#"parsing "+00:00"" as double quoted string: escaped double quote isn't supported"#,
```
Comment (waitingkuo) [inline datafusion/sql/src/expr/arrow_cast.rs:765]: @alamb 
added error cases for 
`empty string` and 
`double quoted string that contains double quote`

File: datafusion/core/tests/sqllogictests/test_files/arrow_typeof.slt [hunk 5]
```diff
@@ -303,3 +307,25 @@ select arrow_cast(interval '30 minutes', 'Duration(Second)');
...
+query P
+select arrow_cast(timestamp '2000-01-01T00:00:00', 'Timestamp(Nanosecond, Some( "+08:00" ))');
+----
+2000-01-01T08:00:00+08:00
+
+# once https://github.com/apache/arrow-rs/issues/1936 fixed, please update this query
+query P
+select arrow_cast(timestamp '2000-01-01T00:00:00Z', 'Timestamp(Nanosecond, Some( "+08:00" ))');
+----
+2000-01-01T08:00:00+08:00
```
Comment (waitingkuo) [inline datafusion/core/tests/sqllogictests/test_files/arrow_typeof.slt:328]: @tustvold thank you
added this case to notify users once the kernel casting issue fixed

File: datafusion/sql/src/expr/arrow_cast.rs [hunk 6]
```diff
@@ -393,6 +418,44 @@ impl<'a> Tokenizer<'a> {
...
+                if let Some(last_c) = self.word.chars().last() {
+                    if last_c != '"' || len < 2 {
+                        return Err(make_error(
+                            self.val,
+                            &format!("parsing {} as double quoted string: last char must be \"", self.word),
+                        ));
+                    }
+                }
+
+                if len == 2 {
```
Comment (alamb) [inline datafusion/sql/src/expr/arrow_cast.rs:435]: 👍  nice error messages

**Review by alamb — APPROVED** [review]
Love it -- thanks again @waitingkuo

**Pull request merged**